src/devices/Pn5180/Pn5180.cs

```diff
@@ -0,0 +1,1496 @@
+            }
+            catch (TimeoutException)
+            {
+                LogInfo.Log($"{nameof(ReadDataFromCard)}: {nameof(TimeoutException)} in {nameof(SpiWriteRead)}", LogLevel.Debug);
+                return false;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Read data from a card.
+        /// </summary>
+        /// <param name="toRead">>The span of byte to read</param>
+        /// <param name="expectedToRead">The expected number of bytes to read</param>
+        /// <returns>True if success. Will return false if the number of bytes to read is not the same as the expected number to read</returns>
+        /// <remarks>Using this function you'll have to manage yourself the possible low level communication protocol.
+        /// This function write directly to the card all the bytes. Please make sure you'll first load specific radio frequence settings,
+        /// detect a card, select it and then send data</remarks>
+        public bool ReadDataFromCard(Span<byte> toRead, int expectedToRead)
+        {
+            var (numBytes, _) = GetNumberOfBytesReceivedAndValidBits();
+            if (numBytes == expectedToRead)
+            {
+                LogInfo.Log($"{nameof(ReadDataFromCard)}: right number of expected bytes to read", LogLevel.Debug);
+                return ReadDataFromCard(toRead);
+            }
+            else if (numBytes > expectedToRead)
+            {
+                LogInfo.Log($"{nameof(ReadDataFromCard)}: wrong number of expected bytes, clearing the cache", LogLevel.Debug);
+                // Clear all
+                ReadDataFromCard(new byte[numBytes]);
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// REad all the data from the card
+        /// </summary>
+        /// <returns>A byte array with all the read elements, null if nothing can be read</returns>
+        /// <remarks>Using this function you'll have to manage yourself the possible low level communication protocol.
+        /// This function write directly to the card all the bytes. Please make sure you'll first load specific radio frequence settings,
+        /// detect a card, select it and then send data</remarks>
+        public byte[] ReadAllDatFromCard()
+        {
+            LogInfo.Log($"{nameof(ReadAllDatFromCard)}: ", LogLevel.Debug);
+            var (numBytes, _) = GetNumberOfBytesReceivedAndValidBits();
+            if (numBytes < 0)
+            {
+                return null;
+            }
+
+            Span<byte> toRead = stackalloc byte[numBytes];
+            var ret = ReadDataFromCard(toRead);
+            if (ret)
+            {
+                return toRead.ToArray();
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Get the number of bytes to read and the valid number of bits in the last byte
+        /// If the full byte is valid then the value of the valid bit is 0
```

Comment:
you might want to mention if validBits==3 means 3 most significant or least significant bits